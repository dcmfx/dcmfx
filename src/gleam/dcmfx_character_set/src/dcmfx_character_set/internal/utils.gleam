import gleam/string

/// Returns the codepoint for the replacement character '�' that is emitted when
/// invalid string data is encountered.
///
pub fn replacement_character() -> UtfCodepoint {
  let assert Ok(codepoint) = string.utf_codepoint(0xFFFD)
  codepoint
}

/// Converts an integer codepoint value to a `UtfCodepoint`. The replacement
/// character '�' is returned if the integer is not a valid codepoint.
///
pub fn int_to_codepoint(i: Int) -> UtfCodepoint {
  case string.utf_codepoint(i) {
    Ok(codepoint) -> codepoint
    Error(Nil) -> replacement_character()
  }
}
