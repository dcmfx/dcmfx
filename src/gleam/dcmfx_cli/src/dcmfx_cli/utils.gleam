import gleam/result
import simplifile

/// Returns whether the specified file paths point to the same file on the same
/// device or volume once symlinks are resolved.
///
pub fn is_same_file(
  file0: String,
  file1: String,
) -> Result(Bool, simplifile.FileError) {
  use file0_info <- result.try(simplifile.file_info(file0))
  use file1_info <- result.try(simplifile.file_info(file1))

  Ok(file0_info == file1_info)
}
