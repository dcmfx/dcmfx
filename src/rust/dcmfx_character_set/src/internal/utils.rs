/// The `char` for the replacement character '�' that is emitted when invalid
/// string data is encountered.
///
pub(crate) const REPLACEMENT_CHARACTER: char = '�';

/// Converts an integer codepoint value to a `char`. The replacement character
/// '�' is returned if the integer is not a valid codepoint.
///
pub(crate) fn codepoint_to_char(codepoint: u32) -> char {
  char::from_u32(codepoint).unwrap_or(REPLACEMENT_CHARACTER)
}
