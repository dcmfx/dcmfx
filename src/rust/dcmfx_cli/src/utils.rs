use std::{
  fs::File,
  io::Write,
  path::{Path, PathBuf},
};

use dcmfx::p10::P10Error;

/// Opens an output stream for the given path, first checking whether it exists
/// and prompting the user about overwriting it if necessary. This prompt isn't
/// presented to the user if `force_overwrite` is true.
///
/// The path "-" is interpreted as writing to stdout.
///
pub fn open_output_stream(
  path: &PathBuf,
  display_path: Option<&PathBuf>,
  force_overwrite: bool,
) -> Result<Box<dyn std::io::Write>, P10Error> {
  if *path == PathBuf::from("-") {
    Ok(Box::new(std::io::stdout()))
  } else {
    if let Some(display_path) = display_path {
      println!("Writing \"{}\" â€¦", display_path.display());
    }

    if !force_overwrite {
      prompt_to_overwrite_if_exists(path);
    }

    match File::create(path) {
      Ok(file) => Ok(Box::new(file)),

      Err(e) => Err(P10Error::FileError {
        when: "Opening file".to_string(),
        details: e.to_string(),
      }),
    }
  }
}

/// Prompts the user about overwriting the given file if it exists.
///
pub fn prompt_to_overwrite_if_exists(path: &Path) {
  if !path.exists() {
    return;
  }

  print!(
    "File \"{}\" already exists. Overwrite? (y/N): ",
    path.display()
  );
  std::io::stdout().flush().unwrap();

  let mut input = String::new();
  std::io::stdin().read_line(&mut input).unwrap();
  let input = input.trim().to_lowercase();

  if input != "y" && input != "yes" {
    std::process::exit(1)
  }
}
