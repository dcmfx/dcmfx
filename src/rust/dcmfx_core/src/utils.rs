#[cfg(feature = "std")]
use std::rc::Rc;

#[cfg(not(feature = "std"))]
use alloc::{format, rc::Rc, string::String, vec, vec::Vec};

/// A reference counted byte slice holds an `Rc<Vec<u8>>` to a shareable buffer,
/// along with a range that specifies the part of that buffer that this slice
/// refers to.
///
/// This type is used widely to avoid copying buffers wherever possible, and in
/// most cases can be used like a `&[u8]` would be.
///
#[derive(Clone)]
pub struct RcByteSlice {
  data: Rc<Vec<u8>>,
  range: core::ops::Range<usize>,
}

impl RcByteSlice {
  /// Creates a new referenced counted byte slice from a `Vec<u8>`.
  ///
  pub fn from_vec(data: Vec<u8>) -> Self {
    let range = 0..data.len();

    Self {
      data: Rc::new(data),
      range,
    }
  }

  /// Creates an empty reference counted byte slice.
  ///
  pub fn empty() -> Self {
    Self {
      data: Rc::new(vec![]),
      range: 0..0,
    }
  }

  /// Slices this referenced counted byte slice, returning a new reference
  /// counted byte slice that points to the same underlying data.
  ///
  /// This function does not copy any data.
  ///
  pub fn slice(&self, start: usize, end: usize) -> Self {
    assert!(start <= end, "Byte slice range out of bounds");
    assert!(end <= self.range.len(), "Byte slice range out of bounds");

    Self {
      data: self.data.clone(),
      range: (self.range.start + start)..(self.range.start + end),
    }
  }

  /// Returns a new reference counted byte slice with the specified number of
  /// bytes dropped from the front.
  ///
  pub fn drop(&self, n: usize) -> Self {
    self.slice(n, self.len())
  }

  /// Returns a new reference counted byte slice to the specified number of
  /// leading bytes.
  ///
  pub fn take(&self, n: usize) -> Self {
    self.slice(0, n)
  }

  /// Consumes this reference counted byte slice and turns it into a `Vec<u8>`.
  /// Avoids a copy when possible.
  ///
  /// This function copies data if there are multiple references to the
  /// underlying buffer, or its slice bounds do not cover the whole buffer.
  ///
  pub fn into_vec(self) -> Vec<u8> {
    if self.range == (0..self.data.len()) {
      match Rc::try_unwrap(self.data) {
        Ok(data) => data,
        Err(data_rc) => data_rc[self.range.clone()].to_vec(),
      }
    } else {
      self.as_slice().to_vec()
    }
  }

  fn as_slice(&self) -> &[u8] {
    &self.data[self.range.clone()]
  }
}

impl core::fmt::Debug for RcByteSlice {
  fn fmt(&self, f: &mut core::fmt::Formatter<'_>) -> core::fmt::Result {
    write!(f, "{:?}", self.as_slice())
  }
}

impl PartialEq for RcByteSlice {
  fn eq(&self, other: &Self) -> bool {
    self.as_slice() == other.as_slice()
  }
}

impl core::ops::Deref for RcByteSlice {
  type Target = [u8];

  fn deref(&self) -> &Self::Target {
    self.as_slice()
  }
}

impl From<Vec<u8>> for RcByteSlice {
  fn from(vec: Vec<u8>) -> Self {
    RcByteSlice::from_vec(vec)
  }
}

/// Inspects a byte slice in hexadecimal, e.g. `[1A 2B 3C 4D]`. If the number of
/// bytes in the slice exceeds `max_length` then not all bytes will be
/// shown and a trailing ellipsis will be appended, e.g. `[1A 2B 3C 4D …]`.
///
pub fn inspect_u8_slice(bytes: &[u8], max_length: usize) -> String {
  let byte_count = core::cmp::min(max_length, bytes.len());

  let s = bytes[0..byte_count]
    .iter()
    .map(|byte| format!("{:02X}", byte))
    .collect::<Vec<_>>()
    .join(" ");

  if byte_count == bytes.len() {
    format!("[{}]", s)
  } else {
    format!("[{} …]", s)
  }
}

#[cfg(test)]
mod tests {
  use super::*;

  #[cfg(not(feature = "std"))]
  use alloc::string::ToString;

  #[test]
  fn inspect_u8_slice_test() {
    assert_eq!(
      inspect_u8_slice(&[0xD1, 0x96, 0x33], 100),
      "[D1 96 33]".to_string()
    );

    assert_eq!(
      inspect_u8_slice(&[0xD1, 0x96, 0x33, 0x44], 3),
      "[D1 96 33 …]".to_string()
    );
  }
}
