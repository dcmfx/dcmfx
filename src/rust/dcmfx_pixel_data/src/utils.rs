use core::ops::{Add, Div};

/// Unsigned integer division that rounds to the nearest integer instead of
/// truncating towards zero like standard integer division.
///
#[inline]
pub fn udiv_round<T>(numer: T, denom: T) -> T
where
  T: Copy + Add<Output = T> + Div<Output = T> + From<u8>,
{
  (numer + (denom / T::from(2))) / denom
}
