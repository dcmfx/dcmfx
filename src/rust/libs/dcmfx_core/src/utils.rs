/// Removes all whitespace from the end of the passed string. Whitespace is
/// defined as the following Unicode codepoints: U+0000, U+0009, U+000A, U+000D,
/// U+0020.
///
pub fn trim_right_whitespace(s: &str) -> &str {
  s.trim_end_matches([
    '\u{0000}', '\u{0009}', '\u{000A}', '\u{000D}', '\u{0020}',
  ])
}
