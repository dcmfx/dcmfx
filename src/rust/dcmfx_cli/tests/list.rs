mod utils;

use assert_cmd::Command;
use insta::assert_snapshot;
use utils::{get_stderr, get_stdout, to_native_path};

#[test]
fn with_multiple_directories() {
  let mut cmd = Command::cargo_bin("dcmfx_cli").unwrap();

  let assert = cmd
    .arg("list")
    .arg("../../../test/assets/fo-dicom")
    .arg("../../../test/assets/pydicom")
    .assert()
    .success();

  let mut lines: Vec<_> =
    get_stdout(assert).split("\n").map(to_native_path).collect();
  lines.sort();

  #[cfg(windows)]
  assert_snapshot!("with_multiple_directories_windows", lines.join("\n"));

  #[cfg(not(windows))]
  assert_snapshot!("with_multiple_directories", lines.join("\n"));
}

#[test]
fn with_summary() {
  let mut cmd = Command::cargo_bin("dcmfx_cli").unwrap();

  let assert = cmd
    .arg("list")
    .arg("../../../test/assets/fo-dicom")
    .arg("--extension")
    .arg("dcm")
    .arg("--summarize")
    .assert()
    .success();

  assert_eq!(
    get_stderr(assert).trim(),
    "Found 32 DICOM files, size: 11.2 MiB"
  );
}

#[test]
fn with_json_lines_format() {
  let mut cmd = Command::cargo_bin("dcmfx_cli").unwrap();

  let assert = cmd
    .arg("list")
    .arg(to_native_path("../../../test/assets/fo-dicom"))
    .arg(to_native_path("../../../test/assets/pydicom"))
    .arg("--format")
    .arg("json-lines")
    .assert()
    .success();

  let stdout = get_stdout(assert);
  let mut lines: Vec<_> = stdout.split("\n").collect();
  lines.sort();

  #[cfg(windows)]
  assert_snapshot!("with_json_lines_format_windows", lines.join("\n"));

  #[cfg(not(windows))]
  assert_snapshot!("with_json_lines_format", lines.join("\n"));
}

#[test]
fn with_invalid_directory() {
  let mut cmd = Command::cargo_bin("dcmfx_cli").unwrap();

  let assert = cmd.arg("list").arg("missing-directory").assert().failure();

  #[cfg(windows)]
  assert_snapshot!("with_invalid_directory_windows", get_stderr(assert));

  #[cfg(not(windows))]
  assert_snapshot!("with_invalid_directory", get_stderr(assert));
}
