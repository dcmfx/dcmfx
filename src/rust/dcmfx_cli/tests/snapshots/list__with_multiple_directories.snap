---
source: dcmfx_cli/tests/list.rs
expression: "lines.join(\"\\n\")"
snapshot_kind: text
---
../../../test/assets/fo-dicom/CR-ModalitySequenceLUT.dcm
../../../test/assets/fo-dicom/CT-MONO2-16-ankle.dcm
../../../test/assets/fo-dicom/CT1_J2KI.dcm
../../../test/assets/fo-dicom/DICOMDIR.dcm
../../../test/assets/fo-dicom/D_CLUNIE_CT1_RLE_FRAGS.dcm
../../../test/assets/fo-dicom/GH064.dcm
../../../test/assets/fo-dicom/GH1146.dcm
../../../test/assets/fo-dicom/GH1261.dcm
../../../test/assets/fo-dicom/GH1308.dcm
../../../test/assets/fo-dicom/GH177_D_CLUNIE_CT1_IVRLE_BigEndian_ELE_undefinded_length.dcm
../../../test/assets/fo-dicom/GH178.dcm
../../../test/assets/fo-dicom/GH184.dcm
../../../test/assets/fo-dicom/GH220.dcm
../../../test/assets/fo-dicom/GH223.dcm
../../../test/assets/fo-dicom/GH342.dcm
../../../test/assets/fo-dicom/GH355.dcm
../../../test/assets/fo-dicom/GH487.dcm
../../../test/assets/fo-dicom/GH538-jpeg1.dcm
../../../test/assets/fo-dicom/GH538-jpeg14sv1.dcm
../../../test/assets/fo-dicom/GH610.dcm
../../../test/assets/fo-dicom/HasFragmentSequence.dcm
../../../test/assets/fo-dicom/IM-0001-0001-0001.dcm
../../../test/assets/fo-dicom/Issue1097_FailToOpenDeflatedFileWithSQ.dcm
../../../test/assets/fo-dicom/JPEGwithIcon.dcm
../../../test/assets/fo-dicom/OutOfBoundsOverlay.dcm
../../../test/assets/fo-dicom/RI6XDosiGap.dcm
../../../test/assets/fo-dicom/TestPattern_Palette.dcm
../../../test/assets/fo-dicom/TestPattern_Palette_16.dcm
../../../test/assets/fo-dicom/TestPattern_RGB.dcm
../../../test/assets/fo-dicom/genFile.dcm
../../../test/assets/fo-dicom/mr_brucker.dcm
../../../test/assets/fo-dicom/multiframe.dcm
../../../test/assets/pydicom/charset_files/chrArab.dcm
../../../test/assets/pydicom/charset_files/chrFren.dcm
../../../test/assets/pydicom/charset_files/chrFrenMulti.dcm
../../../test/assets/pydicom/charset_files/chrGerm.dcm
../../../test/assets/pydicom/charset_files/chrGreek.dcm
../../../test/assets/pydicom/charset_files/chrH31.dcm
../../../test/assets/pydicom/charset_files/chrH32.dcm
../../../test/assets/pydicom/charset_files/chrHbrw.dcm
../../../test/assets/pydicom/charset_files/chrI2.dcm
../../../test/assets/pydicom/charset_files/chrJapMulti.dcm
../../../test/assets/pydicom/charset_files/chrJapMultiExplicitIR6.dcm
../../../test/assets/pydicom/charset_files/chrKoreanMulti.dcm
../../../test/assets/pydicom/charset_files/chrRuss.dcm
../../../test/assets/pydicom/charset_files/chrSQEncoding.dcm
../../../test/assets/pydicom/charset_files/chrSQEncoding1.dcm
../../../test/assets/pydicom/charset_files/chrX1.dcm
../../../test/assets/pydicom/charset_files/chrX2.dcm
../../../test/assets/pydicom/palettes/fall.dcm
../../../test/assets/pydicom/palettes/hotiron.dcm
../../../test/assets/pydicom/palettes/hotmetalblue.dcm
../../../test/assets/pydicom/palettes/pet.dcm
../../../test/assets/pydicom/palettes/pet20step.dcm
../../../test/assets/pydicom/palettes/spring.dcm
../../../test/assets/pydicom/palettes/summer.dcm
../../../test/assets/pydicom/test_files/693_J2KI.dcm
../../../test/assets/pydicom/test_files/CT_small.dcm
../../../test/assets/pydicom/test_files/ExplVR_BigEnd.dcm
../../../test/assets/pydicom/test_files/GDCMJ2K_TextGBR.dcm
../../../test/assets/pydicom/test_files/J2K_pixelrep_mismatch.dcm
../../../test/assets/pydicom/test_files/JPEG-lossy.dcm
../../../test/assets/pydicom/test_files/JPEG2000-embedded-sequence-delimiter.dcm
../../../test/assets/pydicom/test_files/JPEG2000.dcm
../../../test/assets/pydicom/test_files/JPEGLSNearLossless_08.dcm
../../../test/assets/pydicom/test_files/JPEGLSNearLossless_16.dcm
../../../test/assets/pydicom/test_files/JPGExtended.dcm
../../../test/assets/pydicom/test_files/MR_small.dcm
../../../test/assets/pydicom/test_files/MR_small_RLE.dcm
../../../test/assets/pydicom/test_files/MR_small_bigendian.dcm
../../../test/assets/pydicom/test_files/MR_small_expb.dcm
../../../test/assets/pydicom/test_files/MR_small_implicit.dcm
../../../test/assets/pydicom/test_files/MR_small_jp2klossless.dcm
../../../test/assets/pydicom/test_files/MR_small_jpeg_ls_lossless.dcm
../../../test/assets/pydicom/test_files/MR_small_padded.dcm
../../../test/assets/pydicom/test_files/SC_jpeg_no_color_transform.dcm
../../../test/assets/pydicom/test_files/SC_jpeg_no_color_transform_2.dcm
../../../test/assets/pydicom/test_files/SC_rgb_dcmtk_+eb+cr.dcm
../../../test/assets/pydicom/test_files/SC_rgb_dcmtk_+eb+cy+n1.dcm
../../../test/assets/pydicom/test_files/SC_rgb_dcmtk_+eb+cy+n2.dcm
../../../test/assets/pydicom/test_files/SC_rgb_dcmtk_+eb+cy+np.dcm
../../../test/assets/pydicom/test_files/SC_rgb_dcmtk_+eb+cy+s2.dcm
../../../test/assets/pydicom/test_files/SC_rgb_dcmtk_+eb+cy+s4.dcm
../../../test/assets/pydicom/test_files/SC_rgb_gdcm_KY.dcm
../../../test/assets/pydicom/test_files/SC_rgb_jls_lossy_line.dcm
../../../test/assets/pydicom/test_files/SC_rgb_jls_lossy_sample.dcm
../../../test/assets/pydicom/test_files/SC_rgb_jpeg_app14_dcmd.dcm
../../../test/assets/pydicom/test_files/SC_rgb_jpeg_dcmd.dcm
../../../test/assets/pydicom/test_files/SC_rgb_jpeg_dcmtk.dcm
../../../test/assets/pydicom/test_files/SC_rgb_jpeg_gdcm.dcm
../../../test/assets/pydicom/test_files/SC_rgb_jpeg_lossy_gdcm.dcm
../../../test/assets/pydicom/test_files/SC_rgb_rle.dcm
../../../test/assets/pydicom/test_files/SC_rgb_rle_16bit.dcm
../../../test/assets/pydicom/test_files/SC_rgb_rle_16bit_2frame.dcm
../../../test/assets/pydicom/test_files/SC_rgb_rle_2frame.dcm
../../../test/assets/pydicom/test_files/SC_rgb_rle_32bit.dcm
../../../test/assets/pydicom/test_files/SC_rgb_rle_32bit_2frame.dcm
../../../test/assets/pydicom/test_files/SC_rgb_small_odd.dcm
../../../test/assets/pydicom/test_files/SC_rgb_small_odd_big_endian.dcm
../../../test/assets/pydicom/test_files/SC_rgb_small_odd_jpeg.dcm
../../../test/assets/pydicom/test_files/SC_ybr_full_422_uncompressed.dcm
../../../test/assets/pydicom/test_files/UN_sequence.dcm
../../../test/assets/pydicom/test_files/dicomdirtests/77654033/CR1/6154.dcm
../../../test/assets/pydicom/test_files/dicomdirtests/77654033/CR2/6247.dcm
../../../test/assets/pydicom/test_files/dicomdirtests/77654033/CR3/6278.dcm
../../../test/assets/pydicom/test_files/dicomdirtests/77654033/CT2/17106.dcm
../../../test/assets/pydicom/test_files/dicomdirtests/77654033/CT2/17136.dcm
../../../test/assets/pydicom/test_files/dicomdirtests/77654033/CT2/17166.dcm
../../../test/assets/pydicom/test_files/dicomdirtests/77654033/CT2/17196.dcm
../../../test/assets/pydicom/test_files/dicomdirtests/98892001/CT2N/6293.dcm
../../../test/assets/pydicom/test_files/dicomdirtests/98892001/CT2N/6924.dcm
../../../test/assets/pydicom/test_files/dicomdirtests/98892001/CT5N/2062.dcm
../../../test/assets/pydicom/test_files/dicomdirtests/98892001/CT5N/2392.dcm
../../../test/assets/pydicom/test_files/dicomdirtests/98892001/CT5N/2693.dcm
../../../test/assets/pydicom/test_files/dicomdirtests/98892001/CT5N/3023.dcm
../../../test/assets/pydicom/test_files/dicomdirtests/98892001/CT5N/3353.dcm
../../../test/assets/pydicom/test_files/dicomdirtests/98892003/MR1/15820.dcm
../../../test/assets/pydicom/test_files/dicomdirtests/98892003/MR1/4919.dcm
../../../test/assets/pydicom/test_files/dicomdirtests/98892003/MR1/5641.dcm
../../../test/assets/pydicom/test_files/dicomdirtests/98892003/MR2/15970.dcm
../../../test/assets/pydicom/test_files/dicomdirtests/98892003/MR2/4950.dcm
../../../test/assets/pydicom/test_files/dicomdirtests/98892003/MR2/4981.dcm
../../../test/assets/pydicom/test_files/dicomdirtests/98892003/MR2/5011.dcm
../../../test/assets/pydicom/test_files/dicomdirtests/98892003/MR2/6273.dcm
../../../test/assets/pydicom/test_files/dicomdirtests/98892003/MR2/6605.dcm
../../../test/assets/pydicom/test_files/dicomdirtests/98892003/MR2/6935.dcm
../../../test/assets/pydicom/test_files/dicomdirtests/98892003/MR700/4467.dcm
../../../test/assets/pydicom/test_files/dicomdirtests/98892003/MR700/4528.dcm
../../../test/assets/pydicom/test_files/dicomdirtests/98892003/MR700/4558.dcm
../../../test/assets/pydicom/test_files/dicomdirtests/98892003/MR700/4588.dcm
../../../test/assets/pydicom/test_files/dicomdirtests/98892003/MR700/4618.dcm
../../../test/assets/pydicom/test_files/dicomdirtests/98892003/MR700/4648.dcm
../../../test/assets/pydicom/test_files/dicomdirtests/98892003/MR700/4678.dcm
../../../test/assets/pydicom/test_files/dicomdirtests/DICOMDIR-bigEnd.dcm
../../../test/assets/pydicom/test_files/dicomdirtests/DICOMDIR-empty.dcm
../../../test/assets/pydicom/test_files/dicomdirtests/DICOMDIR-implicit.dcm
../../../test/assets/pydicom/test_files/dicomdirtests/DICOMDIR-nooffset.dcm
../../../test/assets/pydicom/test_files/dicomdirtests/DICOMDIR-nopatient.dcm
../../../test/assets/pydicom/test_files/dicomdirtests/DICOMDIR-reordered.dcm
../../../test/assets/pydicom/test_files/dicomdirtests/DICOMDIR.dcm
../../../test/assets/pydicom/test_files/dicomdirtests/TINY_ALPHA/DICOMDIR.dcm
../../../test/assets/pydicom/test_files/dicomdirtests/TINY_ALPHA/PT000000/ST000000/SE000000/IM000000.dcm
../../../test/assets/pydicom/test_files/dicomdirtests/TINY_ALPHA/PT000000/ST000000/SE000000/IM000001.dcm
../../../test/assets/pydicom/test_files/dicomdirtests/TINY_ALPHA/PT000000/ST000000/SE000000/IM000002.dcm
../../../test/assets/pydicom/test_files/dicomdirtests/TINY_ALPHA/PT000000/ST000000/SE000000/IM000003.dcm
../../../test/assets/pydicom/test_files/dicomdirtests/TINY_ALPHA/PT000000/ST000000/SE000000/IM000004.dcm
../../../test/assets/pydicom/test_files/dicomdirtests/TINY_ALPHA/PT000000/ST000000/SE000000/IM000005.dcm
../../../test/assets/pydicom/test_files/dicomdirtests/TINY_ALPHA/PT000000/ST000000/SE000000/IM000006.dcm
../../../test/assets/pydicom/test_files/dicomdirtests/TINY_ALPHA/PT000000/ST000000/SE000000/IM000007.dcm
../../../test/assets/pydicom/test_files/dicomdirtests/TINY_ALPHA/PT000000/ST000000/SE000000/IM000008.dcm
../../../test/assets/pydicom/test_files/dicomdirtests/TINY_ALPHA/PT000000/ST000000/SE000000/IM000009.dcm
../../../test/assets/pydicom/test_files/dicomdirtests/TINY_ALPHA/PT000000/ST000000/SE000000/IM00000A.dcm
../../../test/assets/pydicom/test_files/dicomdirtests/TINY_ALPHA/PT000000/ST000000/SE000000/IM00000B.dcm
../../../test/assets/pydicom/test_files/dicomdirtests/TINY_ALPHA/PT000000/ST000000/SE000000/IM00000C.dcm
../../../test/assets/pydicom/test_files/dicomdirtests/TINY_ALPHA/PT000000/ST000000/SE000000/IM00000D.dcm
../../../test/assets/pydicom/test_files/dicomdirtests/TINY_ALPHA/PT000000/ST000000/SE000000/IM00000E.dcm
../../../test/assets/pydicom/test_files/dicomdirtests/TINY_ALPHA/PT000000/ST000000/SE000000/IM00000F.dcm
../../../test/assets/pydicom/test_files/dicomdirtests/TINY_ALPHA/PT000000/ST000000/SE000000/IM00000G.dcm
../../../test/assets/pydicom/test_files/dicomdirtests/TINY_ALPHA/PT000000/ST000000/SE000000/IM00000H.dcm
../../../test/assets/pydicom/test_files/dicomdirtests/TINY_ALPHA/PT000000/ST000000/SE000000/IM00000I.dcm
../../../test/assets/pydicom/test_files/dicomdirtests/TINY_ALPHA/PT000000/ST000000/SE000000/IM00000J.dcm
../../../test/assets/pydicom/test_files/dicomdirtests/TINY_ALPHA/PT000000/ST000000/SE000000/IM00000K.dcm
../../../test/assets/pydicom/test_files/dicomdirtests/TINY_ALPHA/PT000000/ST000000/SE000000/IM00000L.dcm
../../../test/assets/pydicom/test_files/dicomdirtests/TINY_ALPHA/PT000000/ST000000/SE000000/IM00000M.dcm
../../../test/assets/pydicom/test_files/dicomdirtests/TINY_ALPHA/PT000000/ST000000/SE000000/IM00000N.dcm
../../../test/assets/pydicom/test_files/dicomdirtests/TINY_ALPHA/PT000000/ST000000/SE000000/IM00000O.dcm
../../../test/assets/pydicom/test_files/dicomdirtests/TINY_ALPHA/PT000000/ST000000/SE000000/IM00000P.dcm
../../../test/assets/pydicom/test_files/dicomdirtests/TINY_ALPHA/PT000000/ST000000/SE000000/IM00000Q.dcm
../../../test/assets/pydicom/test_files/dicomdirtests/TINY_ALPHA/PT000000/ST000000/SE000000/IM00000R.dcm
../../../test/assets/pydicom/test_files/dicomdirtests/TINY_ALPHA/PT000000/ST000000/SE000000/IM00000S.dcm
../../../test/assets/pydicom/test_files/dicomdirtests/TINY_ALPHA/PT000000/ST000000/SE000000/IM00000T.dcm
../../../test/assets/pydicom/test_files/dicomdirtests/TINY_ALPHA/PT000000/ST000000/SE000000/IM00000U.dcm
../../../test/assets/pydicom/test_files/dicomdirtests/TINY_ALPHA/PT000000/ST000000/SE000000/IM00000V.dcm
../../../test/assets/pydicom/test_files/dicomdirtests/TINY_ALPHA/PT000000/ST000000/SE000000/IM00000W.dcm
../../../test/assets/pydicom/test_files/dicomdirtests/TINY_ALPHA/PT000000/ST000000/SE000000/IM00000X.dcm
../../../test/assets/pydicom/test_files/dicomdirtests/TINY_ALPHA/PT000000/ST000000/SE000000/IM00000Y.dcm
../../../test/assets/pydicom/test_files/dicomdirtests/TINY_ALPHA/PT000000/ST000000/SE000000/IM00000Z.dcm
../../../test/assets/pydicom/test_files/dicomdirtests/TINY_ALPHA/PT000000/ST000000/SE000000/IM000010.dcm
../../../test/assets/pydicom/test_files/dicomdirtests/TINY_ALPHA/PT000000/ST000000/SE000000/IM000011.dcm
../../../test/assets/pydicom/test_files/dicomdirtests/TINY_ALPHA/PT000000/ST000000/SE000000/IM000012.dcm
../../../test/assets/pydicom/test_files/dicomdirtests/TINY_ALPHA/PT000000/ST000000/SE000000/IM000013.dcm
../../../test/assets/pydicom/test_files/dicomdirtests/TINY_ALPHA/PT000000/ST000000/SE000000/IM000014.dcm
../../../test/assets/pydicom/test_files/dicomdirtests/TINY_ALPHA/PT000000/ST000000/SE000000/IM000015.dcm
../../../test/assets/pydicom/test_files/dicomdirtests/TINY_ALPHA/PT000000/ST000000/SE000000/IM000016.dcm
../../../test/assets/pydicom/test_files/dicomdirtests/TINY_ALPHA/PT000000/ST000000/SE000000/IM000017.dcm
../../../test/assets/pydicom/test_files/dicomdirtests/TINY_ALPHA/PT000000/ST000000/SE000000/IM000018.dcm
../../../test/assets/pydicom/test_files/dicomdirtests/TINY_ALPHA/PT000000/ST000000/SE000000/IM000019.dcm
../../../test/assets/pydicom/test_files/dicomdirtests/TINY_ALPHA/PT000000/ST000000/SE000000/IM00001A.dcm
../../../test/assets/pydicom/test_files/dicomdirtests/TINY_ALPHA/PT000000/ST000000/SE000000/IM00001B.dcm
../../../test/assets/pydicom/test_files/dicomdirtests/TINY_ALPHA/PT000000/ST000000/SE000000/IM00001C.dcm
../../../test/assets/pydicom/test_files/dicomdirtests/TINY_ALPHA/PT000000/ST000000/SE000000/IM00001D.dcm
../../../test/assets/pydicom/test_files/empty_charset_LEI.dcm
../../../test/assets/pydicom/test_files/examples_jpeg2k.dcm
../../../test/assets/pydicom/test_files/examples_overlay.dcm
../../../test/assets/pydicom/test_files/examples_palette.dcm
../../../test/assets/pydicom/test_files/examples_rgb_color.dcm
../../../test/assets/pydicom/test_files/examples_ybr_color.dcm
../../../test/assets/pydicom/test_files/image_dfl.dcm
../../../test/assets/pydicom/test_files/liver_1frame.dcm
../../../test/assets/pydicom/test_files/liver_expb_1frame.dcm
../../../test/assets/pydicom/test_files/liver_nonbyte_aligned.dcm
../../../test/assets/pydicom/test_files/meta_missing_tsyntax.dcm
../../../test/assets/pydicom/test_files/nested_priv_SQ.dcm
../../../test/assets/pydicom/test_files/no_meta_group_length.dcm
../../../test/assets/pydicom/test_files/priv_SQ.dcm
../../../test/assets/pydicom/test_files/reportsi.dcm
../../../test/assets/pydicom/test_files/reportsi_with_empty_number_tags.dcm
../../../test/assets/pydicom/test_files/rtdose.dcm
../../../test/assets/pydicom/test_files/rtdose_1frame.dcm
../../../test/assets/pydicom/test_files/rtdose_expb.dcm
../../../test/assets/pydicom/test_files/rtdose_expb_1frame.dcm
../../../test/assets/pydicom/test_files/rtdose_rle.dcm
../../../test/assets/pydicom/test_files/rtdose_rle_1frame.dcm
../../../test/assets/pydicom/test_files/rtplan.dcm
../../../test/assets/pydicom/test_files/test-SR.dcm
../../../test/assets/pydicom/test_files/waveform_ecg.dcm
